File: C++ Proof of Concept/Steganography/Steganography/hide.cpp
#include "hide.h"
#include "BitManipulation.h"

#include <vector>
#include <iostream>

using namespace std;

bool canHide(unsigned int sizeOfHaystack, unsigned int sizeOfNeedle, int greed) {
	sizeOfNeedle += 4; //Add size of fixed length header
	int spaceAvailableForHiding = (sizeOfHaystack*greed) / 8;
	cout << "Your needle image will take up " << sizeOfNeedle << " bytes." << endl;
	cout << "Your haystack provides " << spaceAvailableForHiding << " bytes of storage with greed set to " << greed << "." << endl;
	return sizeOfNeedle < spaceAvailableForHiding;
}

bool hide(Image &haystack, Image needle, int greed)
{
	cout << "Your needle image is a " << needle.getWidth() << "x" << needle.getHeight() << " image." << endl;

	if (!canHide(haystack.getSize(), needle.getSize(), greed))
		return false;

	unsigned int sizeOfDataToHide = needle.getSize() + 4;
	unsigned char* dataToHide = new unsigned char[sizeOfDataToHide];

	//Create header
	dataToHide[0] = getByte(needle.getWidth(), 2);
	dataToHide[1] = getByte(needle.getWidth(), 3);
	dataToHide[2] = getByte(needle.getHeight(), 2);
	dataToHide[3] = getByte(needle.getHeight(), 3);

	//Copy image data over
	for (int i = 0; i < needle.getData().size(); i++) {
		dataToHide[i + 4] = needle.getData()[i];
	}

	//Hide needle's data in haystack
	int bitsToWrite = sizeOfDataToHide * 8;
	int dataToHideByte = 0;
	int dataToHideBit = 0;
	for (int haystackByte = 0; haystackByte < haystack.getData().size(); haystackByte++) {
		for (int haystackBit = 8 - greed; haystackBit < 8; haystackBit++) {
			if (bitsToWrite > 0) {
				setBit(haystack.getData()[haystackByte], haystackBit, isBitSet(dataToHide[dataToHideByte], dataToHideBit));
				bitsToWrite--;

				dataToHideBit++;
				if (dataToHideBit >= 8) {
					dataToHideBit = 0;
					dataToHideByte++;
				}
			}
		}
	}

	delete[] dataToHide;

	return true;
}

File: JavaScript Proof of Concept/hide.js
function canHide(sizeOfHaystack, sizeOfNeedle, greed) {
    sizeOfNeedle += 4;
    let spaceAvailableForHiding = (sizeOfHaystack * greed) / 8;
    return sizeOfNeedle < spaceAvailableForHiding;
}

function hide(haystack, needle, greed) {
    if (!canHide(haystack.length, needle.length, greed))
        return null;

    let sizeOfNeedleWithHeader = needle.length + 4;
    let needleDataWithHeader = new Uint8ClampedArray(sizeOfNeedleWithHeader);

    let haystackWithNeedleData = new Uint8ClampedArray(haystack.length);

    needleDataWithHeader[0] = getByte(needle.getWidth(), 2);
    needleDataWithHeader[1] = getByte(needle.getWidth(), 3);
    needleDataWithHeader[2] = getByte(needle.getHeight(), 2);
    needleDataWithHeader[3] = getByte(needle.getHeight(), 3);

    for (let i = 0; i < needle.length; i++) {
		needleDataWithHeader[i + 4] = needle[i];
    }

    for (let i = 0; i < haystack.length; i++) {
		haystackWithNeedleData[i] = haystack[i];
    }

    let needleDataBitsToHide = sizeOfNeedleWithHeader * 8;

    let needleWithHeaderByte = 0;
    let needleWithHeaderBit = 0;

    for (let haystackByte = 0; haystackByte < haystack.length; haystackByte++) {
		for (let haystackBit = 8 - greed; haystackBit < 8; haystackBit++) {
			if (needleDataBitsToHide > 0) {
				setBit(haystackWithNeedleData[haystackByte], haystackBit, isBitSet(needleDataWithHeader[needleWithHeaderByte], needleWithHeaderBit));
				needleDataBitsToHide--;

				needleWithHeaderBit++;
				if (needleWithHeaderBit >= 8) {
					needleWithHeaderBit = 0;
					needleWithHeaderByte++;
                }
            }
        }
    }

    return haystackWithNeedleData;
}